homeassistant/components/sensorpush_cloud/api.py

```diff
@@ -0,0 +1,155 @@
+    AuthorizeRequest,
+    Configuration,
+    Samples,
+    SamplesRequest,
+    Sensors,
+    SensorsRequest,
+)
+from sensorpush_api.rest import ApiException
+
+from homeassistant.core import HomeAssistant
+from homeassistant.util import dt as dt_util
+
+from .const import ACCESS_TOKEN_EXPIRATION, LOGGER, REQUEST_RETRIES, REQUEST_TIMEOUT
+
+
+def api_call[**_P, _R](
+    func: Callable[Concatenate[SensorPushCloudApi, _P], _R],
+) -> Callable[Concatenate[SensorPushCloudApi, _P], _R]:
+    """Decorate API calls to handle SensorPush Cloud exceptions."""
+
+    @wraps(func)
+    def _api_call(self: SensorPushCloudApi, *args: _P.args, **kwargs: _P.kwargs) -> _R:
+        LOGGER.debug(f"API call to {func} with args={args}, kwargs={kwargs}")
+        retries: int = 0
+        while True:
+            try:
+                result = func(self, *args, **kwargs)
+            except Exception as e:
+                # Force reauthorization if an exception occurs to avoid
+                # authorization failures after temporary outages.
+                self.deadline = dt_util.now()
+
+                # The SensorPush Cloud API suffers from frequent exceptions;
+                # requests are retried before raising an error.
+                if retries < REQUEST_RETRIES:
+                    retries = retries + 1
+                    continue
+
+                LOGGER.debug(f"API call to {func} failed after {retries} retries")
+                if isinstance(e, ApiException):
+                    # API exceptions provide a JSON-encoded message in the
+                    # body; otherwise, fall back to the general behavior.
+                    try:
+                        data = json.loads(e.body)
+                        raise SensorPushCloudError(data["message"]) from e
+                    except Exception:  # noqa: BLE001
+                        pass
+                raise SensorPushCloudError from e
+            else:
+                LOGGER.debug(f"API call to {func} succeeded after {retries} retries")
+                return result
+
+    return _api_call
+
+
+class SensorPushCloudError(Exception):
+    """An exception occurred when calling the SensorPush Cloud API."""
+
+
+class SensorPushCloudApi:
+    """SensorPush Cloud API class."""
+
+    hass: HomeAssistant
+    email: str
+    password: str
+    configuration: Configuration
+    api: ApiApi
+    deadline: datetime
+    lock: Lock
+
+    def __init__(self, hass: HomeAssistant, email: str, password: str) -> None:
+        """Initialize the SensorPush Cloud API object."""
+        self.hass = hass
+        self.email = email
+        self.password = password
+        # Generated Swagger clients install default logging handlers that
+        # conflict with Home Assistant; remove before making API calls.
+        self.configuration = Configuration()
+        for logger in self.configuration.logger.values():
+            logger.removeHandler(self.configuration.logger_stream_handler)
```

Comment:
Please fix this directly in the code gen tools. It looks hacky and is a workaround, which should be avoided